apps/passport-server/src/services/telegramService.ts

```diff
@@ -0,0 +1,125 @@
+import { EdDSATicketPCD, EdDSATicketPCDPackage } from "@pcd/eddsa-ticket-pcd";
+import { Bot, InlineKeyboard } from "grammy";
+import { fetchTelegramConversation } from "../database/queries/telegram/fetchTelegramConversation";
+import { insertTelegramConversation } from "../database/queries/telegram/insertTelegramConversation";
+import { markTelegramUserAsVerified } from "../database/queries/telegram/updateTelegramConversation";
+import { ApplicationContext } from "../types";
+import { logger } from "../util/logger";
+
+export class TelegramService {
+  private context: ApplicationContext;
+  private bot: Bot;
+  private privateChatId: string;
+
+  public constructor(
+    context: ApplicationContext,
+    bot: Bot,
+    privateChatId: string
+  ) {
+    this.context = context;
+    this.bot = bot;
+    this.privateChatId = privateChatId;
+
+    this.bot.command("start", async (ctx) => {
+      await insertTelegramConversation(this.context.dbPool, {
+        telegram_chat_id: ctx.chat.id,
+        telegram_user_id: ctx.message!.from.id
+      });
+      ctx.reply(
+        "Welcome! 👋\n\nPlease verify your credentials via the Passport app, so we can add you to the correct channels.",
+        {
+          reply_markup: new InlineKeyboard().url(
+            "Verify with Passport 🚀",
+            encodeURI(
+              `https://staging.pcdpass.xyz/#/get-without-proving?request=${JSON.stringify(
+                {
+                  type: "GetWithoutProving",
+                  returnUrl: "http://localhost:3000/verify", // TODO: Prod
+                  pcdType: "eddsa-ticket-pcd"
+                }
+              )}`
+            )
+          )
+        }
+      );
+    });
+
+    bot.start();
+  }
+
+  public async getBotURL(): Promise<string> {
+    const { username } = await this.bot.api.getMe();
+    return `https://t.me/${username}`;
+  }
+
+  public async verifyEdDSATicket(
+    serializedEdDSATicket: string,
+    telegramUserId: number
+  ): Promise<boolean> {
+    let pcd: EdDSATicketPCD;
+
+    try {
+      pcd = await EdDSATicketPCDPackage.deserialize(
+        JSON.parse(serializedEdDSATicket).pcd
+      );
+    } catch (e) {
+      throw new Error(`Deserialization error, ${e}`);
+    }
+    const verified = await EdDSATicketPCDPackage.verify(pcd);
```

Comment:
you probably need to check that 1) this ticket was signed by the server, whose public key you can load from the url `https://api-staging.pcdpass.xyz/issue/eddsa-public-key`, which you should probably pass in via an environment variable, and 2) that the ticket is actually for the particular event that we're gating telegram access to

Reply:
you also need to check that the person requesting approval actually knows the private corresponding to the `attendeeSemaphoreId` that is signed